Add staging and web upload reads

Tianhao-Gu opened the pull request:
@msneddon @MrCreosote 

Could you please review? Thanks! 

PR Target: add web and staging area file upload_reads functionality.

Tianhao-Gu committed:
add staging and web upload_reads

Tianhao-Gu committed:
oops

Tianhao-Gu commented on lib/ReadsUtils/ReadsUtilsImpl.py:
```diff
@@ -651,6 +827,7 @@ def __init__(self, config):
         self.scratch = config['scratch']
         self.callback_url = os.environ['SDK_CALLBACK_URL']
         self.ws_url = config['workspace-url']
+        self.token_user = os.environ['KB_AUTH_TOKEN'].split('client_id=')[1].split('|')[0]
```
I think it leaves a potential issue if we change token format in the future. But I not sure a right way to get current user_name.

MrCreosote replied:
Get the token from ctx['token'] and the user id from ctx['user_id'].

Tianhao-Gu replied:
@MrCreosote Thanks! using ctx['token']. By the way, in my local test, ctx['user_id'] is None. So I am still using parsing user_name from token.

MrCreosote replied:
You can set up ctx any way you want in the tests: https://github.com/kbaseapps/ReadsUtils/blob/master/test/ReadsUtils_server_test.py#L47

The user id is set when running normally: https://github.com/kbaseapps/ReadsUtils/blob/master/lib/ReadsUtils/ReadsUtilsServer.py#L567

MrCreosote replied:
here's an example: https://github.com/kbaseapps/DataFileUtil/blob/master/test/DataFileUtil_server_test.py#L32-L39

Tianhao-Gu committed:
remove tmp folder

Tianhao-Gu committed:
use ctx[token]

MrCreosote commented:
Recording here that Tian and I discussed merging the 3 upload methods in this PR into one upload method:
1) Keep fwd/rev_id and fwd/rev_file as is, but deprecate
2) Add fwd/rev_reads (or something) inputs
3) Add reads_source input that's in [shock, file, web, staging] (or something)
4) The method should pull reads from the appropriate place

Tianhao-Gu committed:
update spec

Tianhao-Gu committed:
modify upload_reads

Tianhao-Gu committed:
fix existing tests

Tianhao-Gu committed:
fix more original tests

Tianhao-Gu commented:
@MrCreosote Could you please give it another review? Thanks!

MrCreosote commented on ReadsUtils.spec:
```diff
@@ -69,6 +69,13 @@ module ReadsUtils {
+        download_type - download type ['Direct Download', 'FTP', 'DropBox', 'Google Drive']
+        - OR - 
+        fwd_staging_file_name - reads data file name in staging area:
+            either single end reads, forward/left reads, or interleaved reads.
```
Clarify the text at the beginning of the documentation here to cover the new options - only covers the previous options currently.

Tianhao-Gu replied:
Could you explain more on this? Sorry, I think I didn't get what you mean.

MrCreosote replied:
See lines 56-60 - they assume only 2 input types are now available instead of 4 (plus 4 download types, so 7 really)

Tianhao-Gu replied:
👍

MrCreosote commented on lib/ReadsUtils/ReadsUtilsImpl.py:
```diff
@@ -82,9 +87,20 @@ def _check_pos(self, num, name):
+        
+        if fwdstaging:
+            fwdfile = self._download_staging_file(params.get('user_id'), fwdstaging)
+
```
A possible issue with downloading the files here and having the rest of the code treat them as if they already existed locally is that the error messages you get back from problems later in the code might not be helpful - it'll just include the file name of the downloaded file, and not provide information regarding how to find the source of the file. If you look at the error handling in the main method below the errors are different based on whether the file source was in shock or from a local file; my feeling is that something similar needs to be done for the alternate file sources.

MrCreosote replied:
My feeling is that all the download code, whether from shock or elsewhere, should be consolidated in one part of the fn (using DFU as appropriate), and any errors thrown should identify the source of the data

MrCreosote replied:
@Tianhao-Gu is this one addressed?

MrCreosote commented on lib/ReadsUtils/ReadsUtilsImpl.py:
```diff
@@ -117,8 +135,16 @@ def _proc_upload_reads_params(self, params):
         if not shock and revid:
             raise ValueError('Cannot specify a reverse reads file in shock ' +
                              'with a local forward reads file')
```
The code should check that <= 1 of the rev* sources are specified, and that if it is specified, it matches the fwd* source.

Tianhao-Gu replied:
👍

MrCreosote commented on lib/ReadsUtils/ReadsUtilsImpl.py:
```diff
@@ -14,6 +14,11 @@
 from numbers import Number
 import six
 import uuid
+import urllib2
```
I personally prefer requests but eh

MrCreosote commented on lib/ReadsUtils/ReadsUtilsImpl.py:
```diff
@@ -642,6 +668,223 @@ def get_fq_stats(self, reads_object, file_path):
+        directory pattern: /data/bulk/user_name/file_name
+
+        """
+        return '/data/bulk/%s/%s' % (token_user, upload_file_name)
```
does the /data/bulk part need to be configurable?

Tianhao-Gu replied:
Currently no. The mounted directory always starts with /data/bulk/

MrCreosote commented on lib/ReadsUtils/ReadsUtilsImpl.py:
```diff
@@ -642,6 +668,223 @@ def get_fq_stats(self, reads_object, file_path):
+        elif download_type == 'FTP':
+            self._download_ftp_link(file_url, copy_file_path)
+        elif download_type == 'Google Drive':
+            self._download_google_drive_link(file_url, copy_file_path)
```
throw an 'unknown download type: {}' (or something) error here

Tianhao-Gu replied:
👍

MrCreosote commented on lib/ReadsUtils/ReadsUtilsImpl.py:
```diff
@@ -642,6 +668,223 @@ def get_fq_stats(self, reads_object, file_path):
+        return copy_file_path
+
+
+    def _download_file(self, download_type, file_url, copy_file_path):
```
This and its subfunctions seem generally useful enough that they should live in DataFileUtils

MrCreosote replied:
Addressing this after the demo is fine - can you just stick in something like #TODO Tian move to DFU

MrCreosote commented on lib/ReadsUtils/ReadsUtilsImpl.py:
```diff
@@ -642,6 +668,223 @@ def get_fq_stats(self, reads_object, file_path):
+
+        # .gz file handler 
+        # TODO: create separate zip file handler for all download types 
+        if self.ftp_file_name.endswith('.gz'):
```
Why is gz only handled for this case?

Tianhao-Gu replied:
Embarrassed. :/ But Chris and Sunita requested a .gz support on this ftp downloader. I think most of the source has .gz file format. Currently they had to download to local and upload.

MrCreosote replied:
Probably want to separate gunzip from the d/l methods, and have it just run if necessary regardless of the source. I think only shock files are automatically uncompressed by DFU. There's a DFU.unpack_file() (or something like that) function you could use, although like I say it'd probably make sense to move this stuff into DFU eventually.

MrCreosote commented on lib/ReadsUtils/ReadsUtilsImpl.py:
```diff
@@ -642,6 +668,223 @@ def get_fq_stats(self, reads_object, file_path):
+        else:
+            force_download_link = file_url.partition('?')[0] + '?raw=1'
+
+        try: online_file = urllib2.urlopen(force_download_link)
```
This part of the fn is identical to the last part of the direct download fn, could make a subfn

Tianhao-Gu replied:
👍

MrCreosote commented on lib/ReadsUtils/ReadsUtilsImpl.py:
```diff
@@ -642,6 +668,223 @@ def get_fq_stats(self, reads_object, file_path):
+        self._check_ftp_connection(self.ftp_user_name, self.ftp_password, self.ftp_domain, self.ftp_file_path, self.ftp_file_name)
+        
+        ftp_connection = ftplib.FTP(self.ftp_domain)
+        ftp_connection.login(self.ftp_user_name, self.ftp_password)
```
Er, do you really want to pass passwords over a cleartext ftp connection?

MrCreosote replied:
Is this one addressed?

MrCreosote commented on lib/ReadsUtils/ReadsUtilsImpl.py:
```diff
@@ -642,6 +668,223 @@ def get_fq_stats(self, reads_object, file_path):
+
+
+    def _download_staging_file(self, token_user, staging_file_name):
+        """
```
It'd be good for this and the other download fns to log what they're doing

Tianhao-Gu replied:
👍

MrCreosote commented on lib/ReadsUtils/ReadsUtilsImpl.py:
```diff
@@ -642,6 +668,223 @@ def get_fq_stats(self, reads_object, file_path):
+        """
+
+        # process ftp credentials 
+        ftp_url_format = re.match(r'ftp://.*:.*@.*/.*', file_url)
```
Is there a lib that'll parse the url for you?

Tianhao-Gu replied:
hmmm. urlparse doesn't do good job on ftp urls. Wasn't able to find any good source online either.

MrCreosote replied:
Fair enough

MrCreosote commented on lib/ReadsUtils/ReadsUtilsImpl.py:
```diff
@@ -642,6 +668,223 @@ def get_fq_stats(self, reads_object, file_path):
+            self.ftp_file_path = file_url.partition('ftp://')[-1].partition('/')[-1].rpartition('/')[0]
+            self.ftp_file_name = re.search('ftp://.*/(.+$)', file_url).group(1)
+
+        self._check_ftp_connection(self.ftp_user_name, self.ftp_password, self.ftp_domain, self.ftp_file_path, self.ftp_file_name)
```
Pretty sure this is too long for pep8 (we're using 100 chars as a limit)

Tianhao-Gu replied:
👍

MrCreosote commented on lib/ReadsUtils/ReadsUtilsImpl.py:
```diff
@@ -642,6 +668,223 @@ def get_fq_stats(self, reads_object, file_path):
+        file_id = file_url.partition('/d/')[-1].partition('/')[0]
+        force_download_link = force_download_link_prefix + file_id
+
+        try: online_file = urllib2.urlopen(force_download_link)
```
repeated code again

Tianhao-Gu replied:
👍

MrCreosote reviewed (changes requested):
Haven't checked the tests yet, will do when current comments are addressed one way or another

Tianhao-Gu committed:
update

Tianhao-Gu committed:
update ReadsUtils.spec

Tianhao-Gu committed:
add unpack_gz_file subfn

Tianhao-Gu commented:
@MrCreosote I think I have addressed all of your concerns. Expect moving download_file to DataFileUtils part. I think it makes sense to move it to DFU. But do it after contractor's meeting?

Tianhao-Gu committed:
consolidate all downloads

Tianhao-Gu committed:
ooops

Tianhao-Gu commented:
@MrCreosote consolidated all downloads into a subfn.

Tianhao-Gu committed:
fix tests

Tianhao-Gu committed:
add todos

Tianhao-Gu committed:
add more catches

Tianhao-Gu committed:
put a warming for FTP

Tianhao-Gu committed:
support only anonymous FTP

Tianhao-Gu committed:
fix FTP tests

Tianhao-Gu committed:
update more tests

Tianhao-Gu committed:
run make

Tianhao-Gu committed:
update sleep time

Tianhao-Gu commented:
@MrCreosote When you have a minute, could you give it another review? Thanks!

MrCreosote commented on ReadsUtils.spec:
```diff
@@ -55,10 +55,20 @@ module ReadsUtils {
+
+        If web files are specified for upload, a download type one of
+        ['Direct Download', 'DropBox', 'FTP', 'Google Drive'] must be specified too. 
+        The downloadable file must be uncompressed (except for FTP, .gz file is acceptable). 
```
Why doesn't the decompression run regardless of source?

Tianhao-Gu replied:
I put a decompression for just FTP because it's just a temp solution for Chris and Sunita's test. Ideally, after contractor's demo, I can do a concrete decompression in FDU.

MrCreosote replied:
OK.

MrCreosote commented on lib/ReadsUtils/ReadsUtilsImpl.py:
```diff
@@ -82,12 +91,27 @@ def _check_pos(self, num, name):
+            fwdid = fwdstaging
+        elif fwdid:
+            reads_source = 'shock'
+            shock = True
```
reads_source and shock are redundant

Tianhao-Gu replied:
👍

MrCreosote commented on lib/ReadsUtils/ReadsUtilsImpl.py:
```diff
@@ -108,17 +132,31 @@ def _proc_upload_reads_params(self, params):
+        if sum(bool(e) for e in [revid, revfile, revurl, revstaging]) > 1:
+            raise ValueError('Cannot specify more than one rev file source')
+
+        if shock and any([revfile, revurl, revstaging]):
```
Just handle shock the same way as the other types - I don't think there's a need to do anything special is there?

Tianhao-Gu replied:
👍

MrCreosote commented on lib/ReadsUtils/ReadsUtilsImpl.py:
```diff
@@ -133,7 +171,7 @@ def _proc_upload_reads_params(self, params):
         else:
             o = self._build_up_reads_data(params, single_end)
-        return o, wsid, name, objid, kbtype, single_end, fwdid, revid, shock
+        return o, wsid, name, objid, kbtype, single_end, fwdid, revid, reads_source
```
This fn is now 82 lines, should probably split into smaller fns for readability

Tianhao-Gu replied:
👍

MrCreosote commented on lib/ReadsUtils/ReadsUtilsImpl.py:
```diff
@@ -642,6 +680,330 @@ def get_fq_stats(self, reads_object, file_path):
+            os.makedirs(dstdir)
+        shutil.copy2(staging_file_path, dstdir)
+        copy_file_path = os.path.join(dstdir, staging_file_path)
+        self.log('Copied staging file from %s to %s' % (staging_file_path, copy_file_path))
```
Just FYI the new improved way is 'Copied {} to {}'.format(v1, v2) but again, meh

MrCreosote commented on lib/ReadsUtils/ReadsUtilsImpl.py:
```diff
@@ -642,6 +680,330 @@ def get_fq_stats(self, reads_object, file_path):
+        elif download_type == 'FTP':
+            self._download_ftp_link(file_url, copy_file_path)
+        elif download_type == 'Google Drive':
+            self._download_google_drive_link(file_url, copy_file_path)
```
Still need to throw an error here

Tianhao-Gu replied:
👍

MrCreosote commented on lib/ReadsUtils/ReadsUtilsImpl.py:
```diff
@@ -642,6 +680,330 @@ def get_fq_stats(self, reads_object, file_path):
+        print "=== WARNING ===\n=== WARNING ===\n"
+        print "FTP IS NOT A SECURIED PROTOCOL\nCREDENTIALS AND FILE CONTENTS WILL BE EXPOSED TO THE PUBLIC"
+        print "\n=== STOP THE PROCESS NOW IF YOU DO NOT WISH TO EXPOSE PRIVATE DATA ==="
+        time.sleep(60)
```
Sorry, I guess we miscommunicated here. I was thinking of just adding a note in the documentation / UI that ftp credentials would be exposed. Since we're not allowing creds any more I don't that's necessary, and definitely not a 60s pause for every job.

Tianhao-Gu replied:
👍

MrCreosote commented on lib/ReadsUtils/ReadsUtilsImpl.py:
```diff
@@ -642,6 +680,330 @@ def get_fq_stats(self, reads_object, file_path):
+        else:
+            self.log('Setting anonymous FTP user_name and password')
+            self.ftp_user_name = 'anonymous'
+            self.ftp_password = 'anonymous@domain.com'
```
Is this supposed to be the literal domain.com or the actual domain parsed out in the next step?

Tianhao-Gu replied:
The domain in password can be anything. The password can be any email format. The actual domain is parsed in the next step.

MrCreosote commented on lib/ReadsUtils/ReadsUtilsImpl.py:
```diff
@@ -642,6 +680,330 @@ def get_fq_stats(self, reads_object, file_path):
+        self.log('Copied FTP file to: %s' % ftp_copy_file_path)
+
+        if self.ftp_file_name.endswith('.gz'):
+            self._unpack_gz_file(copy_file_path)
```
Again, I'm not sure why we don't just move the decompression outside this method and decompress any input that needs it (other than Shock which is handled by DFU).

Also, DFU uses filemagic to detect gzipped files rather than extensions which is probably more robust, so when this is moved to DFU it should do the same thing.

Tianhao-Gu replied:
yup yup. It's a very very temp thing for now. I am not happy this _unpack_gz_file either.

MrCreosote commented on lib/ReadsUtils/ReadsUtilsImpl.py:
```diff
@@ -642,6 +680,330 @@ def get_fq_stats(self, reads_object, file_path):
+            pass
+        else:
+            raise ValueError("Unexpected reads_source value. reads_source: %s" % reads_source)
+
```
So I'd just do the unzip here

MrCreosote commented on lib/ReadsUtils/ReadsUtilsImpl.py:
```diff
@@ -642,6 +680,330 @@ def get_fq_stats(self, reads_object, file_path):
+
+        return returnVal
+
+    def _generate_validation_error_message(self, reads_source, actualpath, fwdpath, revpath, fwdname, revname, fwdid, revid, fwdurl, revurl, fwdstaging, revstaging):
```
pep8 length

MrCreosote replied:
Actually, I just opened the repo up in Eclipse and the pep8 checker shows a lot of pep8 errors. Can you run it through a pep8 checker and fix the issues?

Tianhao-Gu replied:
👍

MrCreosote commented on lib/ReadsUtils/ReadsUtilsImpl.py:
```diff
@@ -14,6 +14,12 @@
+import ftplib
+import re
+import gzip
+import time
```
double import

Tianhao-Gu replied:
👍

MrCreosote commented on lib/ReadsUtils/ReadsUtilsImpl.py:
```diff
@@ -642,6 +680,330 @@ def get_fq_stats(self, reads_object, file_path):
+                    ". REV Path : " + revpath + ".")
+            else:
+                validation_error_message += (" Input Shock ID : " + 
+                                        fwdid + ". File Name : " + fwdname + ".")
```
Missing the file path

Tianhao-Gu replied:
I did
"validation_error_message = "Invalid FASTQ file - Path: " + actualpath + ".""

at the beginning.

MrCreosote replied:
Oh right, you only need the fwd and rev paths for non-interleaved paired end reads.

MrCreosote commented on lib/ReadsUtils/ReadsUtilsImpl.py:
```diff
@@ -642,6 +680,330 @@ def get_fq_stats(self, reads_object, file_path):
+            else:
+                validation_error_message += (" Input Shock ID : " + 
+                                        fwdid + ". File Name : " + fwdname + ".")
+        elif reads_source == 'web':
```
web & staging should include the file path as well

Tianhao-Gu replied:
👍

MrCreosote commented on lib/ReadsUtils/ReadsUtilsImpl.py:
```diff
@@ -642,6 +680,330 @@ def get_fq_stats(self, reads_object, file_path):
+            fwdpath = self._download_staging_file(user_id, fwd)
+            revpath = self._download_staging_file(user_id, rev) if rev else None
+        elif reads_source == 'local':
+            pass
```
set fwdpath and revpath to fwdid and revid here to maintain the same interface as the rest of the types (see below)

Also no need to declare *path above if you do this

Tianhao-Gu replied:
👍

MrCreosote commented on lib/ReadsUtils/ReadsUtilsImpl.py:
```diff
@@ -892,22 +1284,13 @@ def upload_reads(self, ctx, params):
-            elif revpath:
-                validation_error_message += (" Input Files Paths - FWD Path : " +
-                                             fwdpath + ", REV Path : " + revpath + ".")
+            validation_error_message = self._generate_validation_error_message(
```
Why not just pass in ret (and include fwdid and revid in ret) here? Then the whole block starting with if reads_source == 'shock' above isn't needed, I think.

MrCreosote replied:
So it'd be _gen_val_err(reads_source, actualpath, ret) I think

Tianhao-Gu replied:
hmmm I did this because I no longer want to treat fwdid as fwdurl/fwdstaging. I think it's more confusing to me.

MrCreosote replied:
rename it to fwd_location or something - the point is that the combination of fwdid, revid, and reads_source tell you where to get a set of reads - think of it like an object where the api is fwdid and revid and it has different implementations based on the source, which is specified by reads_source. That approach makes the code way less complicated and results in many fewer variables.

Tianhao-Gu replied:
👍

MrCreosote commented on lib/ReadsUtils/ReadsUtilsImpl.py:
```diff
@@ -642,6 +680,330 @@ def get_fq_stats(self, reads_object, file_path):
+            else:
+                validation_error_message += (" Input Staging : " + fwdstaging + ".")
+        elif reads_source == 'local':
+             if revpath:
```
This this is missing an else for if there is not revpath

Tianhao-Gu replied:
For local file path if not revpath, error message will just be

validation_error_message = "Invalid FASTQ file - Path: " + actualpath + "."

MrCreosote replied:
Gotcha

MrCreosote reviewed (changes requested):
So far I haven't been able to get tests to pass, but I suspect that's due to auth issues and not the tests themselves. Investigating. Submitting this review for now, will review test coverage later.

Tianhao-Gu committed:
update

Tianhao-Gu committed:
update

Tianhao-Gu committed:
update

Tianhao-Gu committed:
update _generate_validation_error_message

Tianhao-Gu committed:
add process_rev/fwd_params

Tianhao-Gu committed:
fix tests

Tianhao-Gu committed:
update

Tianhao-Gu committed:
ahhh fix tests

Tianhao-Gu commented:
@MrCreosote Fixed PEP8 and other concerns. Could you please take another look when you have a sec? Thanks Thanks!

sychan commented:
If the module works reasonably, I would suggest we get this out and then batch up additional changes for a subsequent release. We need to get these uploaders into play ASAP.

Tianhao-Gu commented:
@MrCreosote Yea. It happens to me too. I just had to rerun tests couple of times till it all passes.

MrCreosote commented on lib/ReadsUtils/ReadsUtilsImpl.py:
```diff
@@ -14,6 +14,11 @@
+from contextlib import closing
+import ftplib
+import re
+import gzip
```
I still see ~10 pep8 errors, but meh for now

MrCreosote replied:
Still ~10 pep8 errors and a bunch in the tests files. Seems like Tian's and my pep8 checker have different rules

MrCreosote commented on test/ReadsUtils_server_test.py:
```diff
@@ -10,6 +10,8 @@
 from pprint import pprint
 from zipfile import ZipFile
+import ftplib
+from mock import patch
```
Lots of pep8 errors in this file

MrCreosote replied:
Tests passed 1st time

MrCreosote commented on lib/ReadsUtils/ReadsUtilsImpl.py:
```diff
@@ -106,39 +111,92 @@ def _proc_upload_reads_params(self, params):
+        if sum(bool(e) for e in [revid, revfile, revurl, revstaging]) > 1:
+            raise ValueError('Cannot specify more than one rev file source')
+
+        if reads_source == 'shock' and any([revfile, revurl, revstaging]):
```
Again, I'm not sure why shock is treated differently here - seems like this block could just be deleted and the error made the same type as the others? 

Also it seems like the next block isn't tested.

Tianhao-Gu replied:
👍  removed this block.

Next block actually is tested in test_upload_fail_spec_fwd_file_rev_id.

MrCreosote commented on lib/ReadsUtils/ReadsUtilsImpl.py:
```diff
@@ -637,11 +698,359 @@ def validateFASTA(self, ctx, params):
+        fwdname = None
+        revname = None
+        fwdpath = None
+        revpath = None
```
Last 2 are guaranteed to be set below, no need to declare here

Tianhao-Gu replied:
👍 
remove fwdpath. 
revpath is not set when read_source == 'shock' and not revid.

MrCreosote commented on lib/ReadsUtils/ReadsUtilsImpl.py:
```diff
@@ -637,11 +698,359 @@ def validateFASTA(self, ctx, params):
+        if reads_source == 'shock':
+            # Grab files from Shock
+            fwdid = fwd
+            revid = rev
```
Changing the variable names doesn't seem necessary for the few uses below

Tianhao-Gu replied:
👍

MrCreosote commented on lib/ReadsUtils/ReadsUtilsImpl.py:
```diff
@@ -848,66 +1278,45 @@ def upload_reads(self, ctx, params):
-            actualpath = os.path.join(self.scratch, self.get_file_prefix() + '.inter.fastq')
+            actualpath = os.path.join(
+                self.scratch, self.get_file_prefix() + '.inter.fastq')
             self.interleave(None, None, fwdname, fwdid,
```
Unfortunately I just noticed this but interleave has the same error message issues as validation - right now it only will report shock nodes but not the locations of other files

Tianhao-Gu replied:
@MrCreosote since I am logging all of my func activities, can I not touch this function? I am pretty sure users can find the url/staging file on the log.

MrCreosote replied:
making the users parse through logs to figure out why their upload failed seems really user unfriendly to me. I try to put all the info they need to figure out what went wrong in the error.

Tianhao-Gu replied:
👍

MrCreosote commented on lib/ReadsUtils/ReadsUtilsImpl.py:
```diff
@@ -637,11 +698,359 @@ def validateFASTA(self, ctx, params):
+                "The server couldn\'t fulfill the request.\nError code: %s" % e.code)
+        except urllib2.URLError as e:
+            raise ValueError("Failed to reach a server\nReason: %s" % e.reason)
+        else:
```
Should add the url to the error string

Tianhao-Gu replied:
👍

MrCreosote commented on lib/ReadsUtils/ReadsUtilsImpl.py:
```diff
@@ -637,11 +698,359 @@ def validateFASTA(self, ctx, params):
+
+        """
+        # translate dropbox URL for direct download
+        if "?" not in file_url:
```
I don't see a test for this case

Tianhao-Gu replied:
👍

MrCreosote commented on lib/ReadsUtils/ReadsUtilsImpl.py:
```diff
@@ -637,11 +698,359 @@ def validateFASTA(self, ctx, params):
+        self.log('Connecting FTP link: %s' % file_url)
+        ftp_url_format = re.match(r'ftp://.*:.*@.*/.*', file_url)
+        # process ftp credentials
+        if ftp_url_format:
```
I don't see a happy path test for this case

Tianhao-Gu replied:
👍

MrCreosote commented on lib/ReadsUtils/ReadsUtilsImpl.py:
```diff
@@ -637,11 +698,359 @@ def validateFASTA(self, ctx, params):
+                if file_name in ftp.nlst():
+                    pass
+                else:
+                    raise ValueError("File %s does NOT exist in FTP path: %s" % (
```
I don't see a test for this case

Tianhao-Gu replied:
👍

MrCreosote commented on lib/ReadsUtils/ReadsUtilsImpl.py:
```diff
@@ -637,11 +698,359 @@ def validateFASTA(self, ctx, params):
+                validation_error_message += (" Input Shock ID : " +
+                                             fwdsource + ". File Name : " + fwdname + ".")
+        elif reads_source == 'web':
+            if revsource:
```
I don't see a test for this path

Tianhao-Gu replied:
👍

MrCreosote commented on lib/ReadsUtils/ReadsUtilsImpl.py:
```diff
@@ -637,11 +698,359 @@ def validateFASTA(self, ctx, params):
+            else:
+                validation_error_message += (" Input URL : " + fwdsource + ".")
+        elif reads_source == 'staging':
+            if revsource:
```
I don't see a test for this path

Tianhao-Gu replied:
👍

Tianhao-Gu committed:
fix

Tianhao-Gu committed:
fix interleave message

Tianhao-Gu committed:
fix PEP8

Tianhao-Gu commented:
@MrCreosote Fixed interleave error message.

Tianhao-Gu committed:
oops

MrCreosote commented on lib/ReadsUtils/ReadsUtilsImpl.py:
```diff
@@ -106,39 +111,89 @@ def _proc_upload_reads_params(self, params):
+
+        if revid and reads_source != 'shock':
+            raise ValueError('Cannot specify a reverse reads file in shock ' +
+                             'with a local forward reads file')
```
this error message is inaccurate - it may not be a local reads file. Should say missing shock forward path like the others

Tianhao-Gu replied:
👍

MrCreosote commented on lib/ReadsUtils/ReadsUtilsImpl.py:
```diff
@@ -435,7 +493,18 @@ def interleave(self, source_obj_ref, source_obj_name, fwd_shock_filename,
                         error_message += 'Forward Path {}, Reverse Path {}.'
                         error_message_bindings.extend([fwdpath, revpath])
-                        raise ValueError(error_message.format(*error_message_bindings))
+                        
```
The _read_fq_record funtion called above has the same issue with the errors only mentioning shock and local files

Tianhao-Gu replied:
👍

MrCreosote commented on lib/ReadsUtils/ReadsUtilsImpl.py:
```diff
@@ -435,7 +493,18 @@ def interleave(self, source_obj_ref, source_obj_name, fwd_shock_filename,
+                            error_message += 'Forward File URL {}, Reverse File URL {}.'
+                            error_message_bindings.extend([fwdsource, revsource])
+
+                        if reads_source == 'staging':
```
I don't see a test for this path

Tianhao-Gu replied:
yea. I am aware of this. But I need to mock _get_staging_file_path return. And for this function, _get_staging_file_path needs to return 2 different path. I cannot test it properly.

MrCreosote replied:
Could you just set STAGING_FILE_PREFIX to something else prior to the test, or mock a fn that just returns the value of STAGING_FILE_PREFIX which is then used in _get_s_f?

Tianhao-Gu replied:
👍

MrCreosote commented on lib/ReadsUtils/ReadsUtilsImpl.py:
```diff
@@ -14,6 +14,11 @@
 from numbers import Number
 import six
 import uuid
+import urllib2
```
Oh, forgot - can you bump the version in kbase.yml

Tianhao-Gu replied:
👍

MrCreosote reviewed (changes requested):
Almost there. Forget about the pep8 changes if your checker says everything's fine, we'll figure out the differences later (if we get to it

Tianhao-Gu committed:
udpate

Tianhao-Gu committed:
update

Tianhao-Gu commented:
@MrCreosote Addressed your concerns. Expect one missing test that I couldn't mock _get_staging_file_path to return different values.

MrCreosote commented on test/ReadsUtils_server_test.py:
```diff
@@ -2779,7 +2809,7 @@ def test_upload_reads_from_web_direct_download_paired_ends(self):
         self.assertEqual(d['insert_size_mean'], None)
         self.assertEqual(d['insert_size_std_dev'], None)
-        self.check_lib(d['lib1'], 1341276, file_name, '04dac97530faa081c966e79af7c90883')
+        self.check_lib(d['lib1'], 1341276, file_name, '1c58d7d59c656db39cedcb431376514b')
```
Why did the md5s change? Looking at the code changes I can't figure it out.

Tianhao-Gu replied:
@MrCreosote It was a bug that I found out. 

previously when I called _download_web_file for rev file, I didn't pass rev_file=True argument.

Tianhao-Gu committed:
fix fix

Tianhao-Gu commented:
@MrCreosote YaY Fixed all

MrCreosote commented:
I think it would be better code if there were different class implementations of a common interface for the input types that abstracted away the input type behavior including generating error strings, but that can happen (much) later.

Tests pass.
